src/pip/_internal/req/req_uninstall.py

```diff
@@ -398,6 +398,10 @@ def _display(msg: str, paths: Iterable[str]) -> None:
             will_remove = set(self._paths)
             will_skip = set()
 
+        # git-bash breaks the stdin pipe, so assume confirmation.
+        if os.name == "nt" and os.getenv("MSYSTEM") == "MINGW64":
+            logger.warning("git bash can not prompt input, skipping.")
+            return True  # fixes crash on git-bash win32.
```

Comment:
It seems like it is a bug in git-bash to not allow input to a command. So I’m -1 on having special case code for that. If the `input()` call fails with an exception I would be fine with catching that exception and handling it though.

As for assuming confirmation, IMO that is the wrong default. We want the user to confirm, so if they cannot, we should assume “no” for safety. If they want to say “yes” but cannot respond to the prompt, they should supply `-y`, just like in the case where the command is used in an automated script.

Reply:
> so if they cannot, we should assume “no” for safety.
 
 I don't think assuming -n on exception is an acceptable user experience. In fact, swallowing the exception and assuming n is arguably worse than the status quo, because at least now we get a stack trace when it aborts.

And for what? It's an annoying guardrail anyway. Do you ever recall a single time you've ever hit no to abort a pip uninstall?

Reply:
The better fix would be to figure out why sys.stderr pipe is breaking though...

Reply:
Agreed, we shouldn't have to cater for `sys.stderr` failing.

As to your other point, you're effectively arguing that there's no point prompting at all (after all, you're suggesting that git-bash users will never need to say "no" - what's special about those users?)

> It's an annoying guardrail anyway.

If you believe that, propose that we remove it altogether, don't just remove without prompting if there's an error.

> In fact, swallowing the exception and assuming -n is arguably worse than the status quo, because at least now we get a stack trace when it aborts.

Well, yes. I was assuming that like with any other exception, we'd report the error. (And the error message would suggest using `-y` if that's what the user intended). But to report the error properly, we need to know why git-bash broke `sys.stderr`, which brings us back to the fact that this is likely an error there.

Maybe we could issue an error like "Environment Issue: There is an issue with your shell, as pip is unable to interact with the user. If you want to force uninstallation without prompting, you can use the `-y` flag."

But to be blunt, if we can't write to `sys.stderr`, then we've got far worse problems (for instance, writing the error message I suggested there will fail as well!)

Reply:
FWIW, I don't know how relevant this might be, but I just tested my copy of Python in git-bash (64-bit Windows Python 3.11 downloaded from python.org), just using `py -c "input()"` and `input()` works there. So it looks like there may be a deeper issue here. Can you reproducethe `input()` issue outside of pip? Also, can you show the output of `pip debug` as it might help identify details about your Python interpreter.

Reply:
Are you trying to do this within a virtual environment?

Here are some repro steps that just worked for me on Win32

Start off in `git-bash`

  1. `git clone https://github.com/zackees/transcribe-anything`
  2. `cd transcribe-anything`
  3. `python make_venv.py`
  4. `. activate.sh`
  5.  `pip install -r requirements.txt`
  6. ERROR STEP: `pip uninstall whisper`

The exact commit hash I'm using with `transcribe-anything` is:

8bb939b3f685dcfc6c26484b17b66aedaa54a082

Reply:
Confirmed that this only happens in a virtual environment. In the global python path this does not happen at all.

Reply:
> If you believe that, propose that we remove it altogether, don't just remove without prompting if there's an error.

I believe that this would have been the right move a decade ago. Now there are enough people depending on this behavior that I'm not sure it's the right move anymore. It could turn into a bike shedding moment. Unless there is an indication that people would be in favor of it.

What if we catch the exception and return -y as a fallback behavior? That way we don't have to get into the debate of whether to change the pip uninstall prompting for all platforms, but the git-bash problem is solved. And if someone fixes the stderr pipe situation then the code path will no longer bit hit?

Reply:
I have no `python` executable on my path in git-bash, just the standard `py` launcher that's installed with python.org python. Your reproducer doesn't work, as I get the error `ERROR: No matching distribution found for torch`. Please simplify the reproducer to a single install+uninstall of a package with no dependencies. If you can't do that, it would suggest that the problem is in something other than pip.

Also, as I requested, please provide the output of `pip debug`, and confirm exactly where your python executable was obtained from.